### PR TITLE
DittoFactoryAttribute

### PR DESCRIPTION
**Pull request by mattbrailsford**
Created a DittoFactory processor attribute to allow the dynamic typing of items based upon properties of that item, such as Nested Content items which support a base interface, but may have differing concentrate models named after their DocType alias.

I think more unit tests are needed, and I'm not 100% certain in my head if this is entirely needed, but I put together a POC anyway.

This would be a fix for issue #168 if it was merged in.

### CODE REVIEW

File: src/Our.Umbraco.Ditto/ComponentModel/Processors/DittoFactoryAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            var propType = Context.PropertyDescriptor.PropertyType;
+            var propTypeIsEnumerable = propType.IsEnumerableType();
+            var baseType = propTypeIsEnumerable ? propType.GetEnumerableType() : propType;
+
+            // TODO: Validate the base type more?
+
+            // Find the appropreate types
+            // There is no non generic version of ResolveTypes so we have to
+            // call it via reflection
+            var method = typeof(PluginManager).GetMethod("ResolveTypes");
```
Comment (JimBobSquarePants) [inline src/Our.Umbraco.Ditto/ComponentModel/Processors/DittoFactoryAttribute.cs]: We should save this as a compiled delegate to save looking it up every time.

File: src/Our.Umbraco.Ditto/ComponentModel/Processors/DittoFactoryAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            if (enumerableValue != null)
+            {
+                var items = enumerableValue.Select(x =>
+                {
+                    var typeName = typeNameResolver.ResolveTypeName(Context, x);
+                    var type = types.FirstOrDefault(y => y.Name.InvariantEquals(typeName));
+
+                    // QUESTION: Should we return null? throw an exception? or strip items if a model can't be found?
+
+                    return type != null ? x.As(type): null;
```
Comment (JimBobSquarePants) [inline src/Our.Umbraco.Ditto/ComponentModel/Processors/DittoFactoryAttribute.cs]: I vote null rather than strip.

**Comment by JimBobSquarePants** [comment]
Will need to download and test the code but it looks good to me.

File: src/Our.Umbraco.Ditto/ComponentModel/Processors/DittoFactoryAttribute.cs (hunk 1 above)
Comment (mattbrailsford) [inline src/Our.Umbraco.Ditto/ComponentModel/Processors/DittoFactoryAttribute.cs]: Feel free to weak as I'm not entirely sure how you do those.

Comment (JimBobSquarePants) [inline src/Our.Umbraco.Ditto/ComponentModel/Processors/DittoFactoryAttribute.cs]: magic :wink:

Comment (mattbrailsford) [inline src/Our.Umbraco.Ditto/ComponentModel/Processors/DittoFactoryAttribute.cs]: I have no doubt :8ball:

**Comment by mattbrailsford** [comment]
Wondering if I should drop the use of DittoFactoryTypeNameResolver and instead just make DittoFactoryAttribute abstract and have an abstract method on this called ResolveTypeName(IPublishedContent currentContent)?

It does mean that you couldn't use DittoFactory directly as an attribute, passing a type, but it would mean you can pass additional config in, for example, with the DittoDocTypeFactory, you might want to accept a "Prefix" / "Suffix" parameters so you can easily append something extra on the doctype alias? And I don't see this being any less / more work as you would have had to have defined a custom DittoFactoryTypeNameResolver anywho.

Thoughts?

**Comment by Nicholas-Westby** [comment]
Would love to test this out in a starter kit I'm building for my company. What would be the easiest way of going about that?

Should I clone the repo, switch to the `feature/ditto-factory` branch, and then build the code somehow (I'm guessing you have some streamlined way to do this)? Does building generate binaries or a NuGet package?

**Comment by mattbrailsford** [comment]
You can grab the latest build from the build server
https://ci.appveyor.com/project/leekelleher/umbraco-ditto/build/0.10.0.521/artifacts

**Comment by mattbrailsford** [comment]
Have dropped the DittoFactoryTypeNameResolvers and just gone with an abstract DittoFactoryAttribute instead. Less moving parts, and more options.

**Comment by leekelleher** [comment]
@mattbrailsford All cool with the class name change.

**Comment by leekelleher** [comment]
I've spent some time today going over the codebase, (fixing up StyleCop stuff :smile_cat:; refactoring some unit-tests, etc).  I'm totally happy with this PR being merged in and we can make further unit-tests from the "develop" branch.

@JimBobSquarePants @mattbrailsford Does that sound okay?

**Comment by mattbrailsford** [comment]
Cool with me

**Comment by JimBobSquarePants** [comment]
YASSSS STYLEKITTEN IS A THING NOW! :smile_cat: 

Ahem... Yeah good to go.

**Pull request merged**